docs/book/src/reference/using-finalizers.md

```diff
@@ -0,0 +1,99 @@
+an API type you are implementing and you need to delete the associated external resource
+when the corresponding object is deleted from Kubernetes, you can use a
+finalizer to delete the external resource.
+
+You can read more about the finalizers in the [kubernetes reference docs](https://kubernetes.io/docs/tasks/access-kubernetes-api/custom-resources/custom-resource-definitions/#finalizers). Section below
+demonstrates how to register and trigger pre-delete hooks in the `Reconcile`
+method of a controller.
+
+Highlights:
+- If object is not being deleted and does not have the finalizer registered,
+  then add the finalizer and update the object in kubernetes.
+- If object is being deleted and the finalizer is still present in finalizers list,
+  then execute the pre-delete logic and remove the finalizer and update the
+  object.
+- You should implement the pre-delete logic in such a way that it is safe to 
+ invoke it multiple times for the same object.
+
+```go
+
+func (r *Reconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx := context.Background()
+
+	err := r.Get(ctx, req.NamespacedName, instance)
+	if err != nil {
+		// handle error
+	}
+
+	// name of our custom finalizer
+	myFinalizerName := "storage.finalizers.example.com"
+
+	// examine DeletionTimestamp to determine if object is under deletion
+	if instance.ObjectMeta.DeletionTimestamp.IsZero() {
+		// The object is not being deleted, so if it does not have our finalizer,
+		// then lets add the finalizer and update the object. This is equivalent
+		// registering our finalizer.
+		if !containsString(instance.ObjectMeta.Finalizers, myFinalizerName) {
+			instance.ObjectMeta.Finalizers = append(instance.ObjectMeta.Finalizers, myFinalizerName)
+			if err := r.Update(context.Background(), instance); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+	} else {
+		// The object is being deleted
+		if containsString(instance.ObjectMeta.Finalizers, myFinalizerName) {
+			// our finalizer is present, so lets handle any external dependency
+			if err := r.deleteExternalResources(instance); err != nil {
+				// if fail to delete the external dependency here, return with error
+				// so that it can be retried
+				return ctrl.Result{}, err
+			}
+
+			// remove our finalizer from the list and update it.
+			instance.ObjectMeta.Finalizers = removeString(instance.ObjectMeta.Finalizers, myFinalizerName)
+			if err := r.Update(context.Background(), instance); err != nil {
+				return ctrl.Result{}, err
+			}
+		}
+
+		return ctrl.Result{}, err
+	}
+
+	// rest of the reconciler code
+}
+
+func (r *Reconciler) deleteExternalResources(instance *MyType) error {
+	//
+	// delete any external resources associated with the instance
+	//
+	// Ensure that delete implementation is idempotent and safe to invoke
+	// multiple types for same object.
+}
+
+//
+// Helper functions to check and remove string from a slice of strings.
+//
+func containsString(slice []string, s string) bool {
```

Comment:
We should have a helper for this in CR -- something like `NeedsFinalizing(string, runtime.Object)` and `RemoveFinalizer(string, runtime.Object)`.  Perhaps even wrap those up into: `FinalizeIfNeeded(string, runtime.Object, func())`

Reply:
Agreed. Have created an issue in CR --> https://github.com/kubernetes-sigs/controller-runtime/issues/485